### PR TITLE
Fix handling output from tty-enabled containers.

### PR DESCRIPTION
**Pull request by dano**
As described in issue #630, getting the stdout/stderr from tty-enabled containers doesn't work properly, because docker-py treats the output as multiplexed, when docker is actually sending a raw stream. This patch attempts to address this by determining if a container to is tty-enabled before trying to read its output, and if it is, read the raw stream, rather than reading at a multiplexed stream.

One piece here I'm not 100% sure about is what to do with the logic that was in `attach` to handle old API versions. The logic being used there - read the stream line by line, and skip empty new-lines, doesn't work if the container ever writes blank lines. I also haven't seen any keep-alive new lines being written in my testing, so I wonder if this check isn't really needed?

I suppose I could split `_stream_raw_result` into two methods - one for TTY-enabled containers, and one for versions below 1.6. The one for versions below 1.6 could keep the old, line-by-line behavior.

Also please let me know if any additional unit tests are needed.

### CODE REVIEW

File: docker/client.py [hunk 1]
```diff
@@ -591,15 +590,19 @@ def exec_start(self, exec_id, detach=False, tty=False, stream=False):
...
                               data=data, stream=stream)
+        # Stream multi-plexing was only introduced in API v1.6. Anything before
+        # that needs old-style streaming. We also should use old-style
+        # streaming if we're dealing with a tty-enabled container.
+        if utils.compare_version('1.6', self._version) < 0 or tty:
+            return self._stream_raw_result(res) if stream else \
+                self._result(res, binary=True)
+
         self._raise_for_status(res)
+        sep = bytes() if six.PY3 else str()
```
Comment (aanand) [inline docker/client.py]: This can be replaced with just `six.binary_type()`, I believe.

File: docker/client.py [hunk 2]
```diff
@@ -824,14 +827,21 @@ def logs(self, container, stdout=True, stderr=True, stream=False,
                 params['tail'] = tail
             url = self._url("/containers/{0}/logs".format(container))
             res = self._get(url, params=params, stream=stream)
```
Comment (aanand) [inline docker/client.py]: Can we reduce the duplication here? This is very similar to the same code in `attach()`.

**Comment by aanand** [comment]
> I suppose I could split `_stream_raw_result` into two methods - one for TTY-enabled containers, and one for versions below 1.6. The one for versions below 1.6 could keep the old, line-by-line behavior.

Sounds sensible to me.

> Also please let me know if any additional unit tests are needed.

I'm not sure I fully understand your changes to the tests. Are they now exclusively testing the `Tty=True` case? If so, we should make sure to also test the `Tty=False` case.

**Comment by dano** [comment]
@aanand The tests currently don't actually specify a tty setting - the changes I made add a `tty=False` setting to the mocked container object used in each test, just to keep the tests from failing because `container['Config']['Tty'] is missing. I'll look into adding a test for the `tty=True` case in addition to just keeping the existing tests working.

**Comment by dano** [comment]
Ugh, it looks like my attempt to squash my commits ended up screwing up the branch completely. I'll open a new request...

**Comment by dano** [comment]
Ok, I've opened https://github.com/docker/docker-py/pull/669 to replace this.

**Pull request closed without merging**